test/test_main.py

```diff
@@ -612,10 +661,10 @@ def test_empty_actions(self):
         self._simulate_event(EventSpec(ops.InstallEvent, 'install'))
 
     def test_collect_metrics(self):
-        fake_script(self, 'add-metric', 'exit 0')
+        fake_script(typing.cast(unittest.TestCase, self), 'add-metric', 'exit 0')
```

Comment:
To avoid these annoying `typing.cast` calls (and the list of duplicated assert methods above), can we make `_TestMain` inherit from `unittest.TestCase`, and the subclasses not inherit?

Reply:
Test bother me too :(. `_TestMain` can't inherit from `TestCase` because it's an ABC. Lots of tests start failing like:

```
ERROR test/test_main.py::_TestMain::test_no_actions - TypeError: Can't instantiate abstract class _TestMain with abstract methods _call_event, _setup_entry_point, test_set...
```

It needs to be an ABC because it has tests that can't run without context that the subclasses are providing.

It wouldn't help with the duplicated assert methods, but we could adjust the type of the `test_helpers` methods, for example:

```diff
diff --git a/test/test_helpers.py b/test/test_helpers.py
index ada79af..95f8f8f 100755
--- a/test/test_helpers.py
+++ b/test/test_helpers.py
@@ -25,7 +25,8 @@ from ops.model import _ModelBackend
 from ops.storage import SQLiteStorage
 
 
-def fake_script(test_case: unittest.TestCase, name: str, content: str):
+def fake_script(test_case: typing.Any, name: str, content: str):
+    assert isinstance(test_case, unittest.TestCase)
     if not hasattr(test_case, 'fake_script_path'):
         fake_script_path = tempfile.mkdtemp('-fake_script')
         old_path = os.environ["PATH"]
```

This feels wrong to me - the caller should be making sure the type is correct - but maybe practicality beats purity since this would be one assert vs a whole heap of casts?

Alternatively, there's [making the script functionality part of the class via a mixin, which I did as an exercise to see how large it would be when doing the test_helpers typing](https://github.com/tonyandrewmeyer/operator/commit/46238eee2d0b39d383f2ea6b3d6cac722be5dddd). This is a much bigger diff but avoids all the casting needs.

One other possibility would be that I could make a protocol that defined what `fake_script` and `fake_script_calls` needs and make that the type - I think this could be done so that `TestCase` would match, as well as `_TestMain`. This seems complex for what it's doing, though :(.

Thoughts?

Reply:
Given all of the above, I think what you have now makes the most sense. It's only an issue for tests that do this kind of subclassing, not tests classes that inherit directly from `TestCase`, so it probably doesn't affect *most* usages.

At some point maybe we can clean some of this stuff up and lean into pytest more and use a fixture. But this is fine for now.

Thanks for your efforts on testing the mixing idea though!